torch/utils/data/datapipes/iter/grouping.py

```diff
@@ -43,16 +50,16 @@ def __iter__(self) -> Iterator[List[T_co]]:
             batch.append(x)
             if len(batch) == self.batch_size:
                 yield batch
-                batch.clear()
+                batch = []
         if len(batch) > 0:
             if not self.drop_last:
                 yield batch
-            batch.clear()
+            batch = []
 
     def __len__(self) -> int:
         if self.length is not None:
             return self.length
-        if isinstance(self.datapipe, Sized):
+        if isinstance(self.datapipe, Sized) and self.unbatch_level == 0:
```

Comment:
I don't think we need to check `self.unbatch_level == 0` since `unbatch` doesn't have `__len__` .
https://github.com/pytorch/pytorch/blob/fa030d1213600fa8626540f1b8857cbcbfeeb67c/torch/utils/data/datapipes/iter/grouping.py#L65

Reply:
This could work as optimization, imagine MP situation, when we need to send a message to request len of another pipe (and get special NotAvailable message back),
this shortcircuit will allow skipping this step.